sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -0,0 +1,378 @@
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.coders.RowCoder;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * {@link Schema} describes the fields in {@link Row}.
+ *
+ */
+@Experimental
+@AutoValue
+public abstract class Schema implements Serializable {
+  // A mapping between field names an indices.
+  private BiMap<String, Integer> fieldIndices = HashBiMap.create();
+  public abstract List<Field> getFields();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setFields(List<Field> fields);
+    abstract Schema build();
+  }
+
+  public static Schema of(List<Field> fields) {
+    return Schema.fromFields(fields);
+  }
+
+  public static Schema of(Field ... fields) {
+    return Schema.of(Arrays.asList(fields));
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof Schema)) {
+      return false;
+    }
+    Schema other = (Schema) o;
+    return Objects.equals(fieldIndices, other.fieldIndices)
+        && Objects.equals(getFields(), other.getFields());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(fieldIndices, getFields());
+  }
+
+  /**
+   * An enumerated list of supported types.
+   */
+  public enum TypeName {
+    BYTE,    // One-byte signed integer.
+    INT16,   // two-byte signed integer.
+    INT32,   // four-byte signed integer.
+    INT64,   // eight-byte signed integer.
+    DECIMAL,  // Decimal integer
+    FLOAT,
+    DOUBLE,
+    STRING,  // String.
+    DATETIME, // Date and time.
+    BOOLEAN,  // Boolean.
+    ARRAY,
+    ROW;    // The field is itself a nested row.
+
+    public static final List<TypeName> NUMERIC_TYPES = ImmutableList.of(
+        BYTE, INT16, INT32, INT64, DECIMAL, FLOAT, DOUBLE);
+    public static final List<TypeName> STRING_TYPES = ImmutableList.of(STRING);
+    public static final List<TypeName> DATE_TYPES = ImmutableList.of(DATETIME);
+    public static final List<TypeName> CONTAINER_TYPES = ImmutableList.of(ARRAY);
+    public static final List<TypeName> COMPOSITE_TYPES = ImmutableList.of(ROW);
+
+    public boolean isNumericType() {
+      return NUMERIC_TYPES.contains(this);
+    }
+    public boolean isStringType() {
+      return STRING_TYPES.contains(this);
+    }
+    public boolean isDateType() {
+      return DATE_TYPES.contains(this);
+    }
+    public boolean isContainerType() {
+      return CONTAINER_TYPES.contains(this);
+    }
+    public boolean isCompositeType() {
+      return COMPOSITE_TYPES.contains(this);
+    }
+
+    /** Returns a {@link FieldTypeDescriptor} representing this primitive type. */
+    public FieldTypeDescriptor typeDescriptor() {
+      return FieldTypeDescriptor.of(this);
```

Comment:
Should we cache these? They're immutable, so it probably makes sense to keep an instance in the field in the `TypeName`